### ISSUE TITLE
How to run existing Appium testcases on Perfecto?

### ISSUE THREAD

**Issue by Georgy-Ryabov** [qmetry/qaf#158]
### QAF Version
		<qaf.version>2.1.12</qaf.version>
		<qaf-support.version>2.1.11</qaf-support.version>

Following the document ["How to run existing Appium testcases on Perfecto?"](https://qmetry.github.io/qaf/latest/how_to_run_appium_testcases_on_perfecto.html)

### Preconditions

appium nodeJS is installed.

### Steps To Reproduce

    My configuration (as described in document)

    remote.server=https://mycloud.perfectomobile.com/nexperience/perfectomobile/wd/hub
    env.baseurl=https://mycloud.perfectomobile.com/nexperience/perfectomobile/wd/hub    
    driver.name=appiumDriver
    appName = Contact Manager

    appium.additional.capabilities = {\
      'app':"${app}",\
      'appName': '${appName}',\
      'deviceName': '${deviceName}',\
      'model':'${model}',\
      'manufacturer':'${manufacturer}',\
      'location':'${location}',\
      'resolution':'${resolution}',\
      'platformName':"${platform.name}",\
      'platformVersion':"${platform.version}",\
      'driver.class':"${driver.class}",\
      'sendKeyStrategy':"${send.key.strategy}",\
      'user': '${user}',\
      'password'='${password}',\
      'appWaitActivity':'com.x.w.zActivity',\
      'javascriptEnabled': true,\
      'automationName':'Appium',\
      'noReset': true,\
    }

From the test I try to get driver:
` return (AppiumDriver) new WebDriverTestBase().getDriver().getUnderLayingDriver();`

    When I run test with this configuration I get driver that connected to local service. 

How is it happened? 

in function `QAFExtendedWebDriver.getDriver` I see the next:
```

QAFExtendedWebDriver driver =
				b.contains("remote") 
				        ?  browser.getDriver(urlStr, reporter)
					:  browser.getDriver(reporter, urlStr);
```
b = 'appiumDriver' (see function arg argument, that creates from driver.name=appiumDriver )

After some statements will be called `getDriverObj(driverCls, desiredCapabilities, urlstr);`

Go into the function and see : 
```
			Constructor<? extends WebDriver> constructor =
					of.getConstructor(Capabilities.class);
			return constructor.newInstance(capabilities);
```
>>>
`IOSDriver(Capabilities desiredCapabilities)`
>>>
```
public AppiumDriver(Capabilities desiredCapabilities,
        Class<? extends JsonToWebElementConverter> converterClazz) {
        this(AppiumDriverLocalService.buildDefaultService(), desiredCapabilities, converterClazz);
    }
```
>>> 
`AppiumDriverLocalService.buildDefaultService()`

That creates **local** service using installed nodejs. 
In my case it is /usr/local/Cellar/node@6/6.11.5/bin/node

The property remote.server=https://mycloud.perfectomobile.com/nexperience/perfectomobile/wd/hub is ignored
 

Ok then, I changed my  driver.name=appiumDriver to  driver.name=appium**Remote**Driver

Driver is created and can access to Perfecto, but 
`return (AppiumDriver) new WebDriverTestBase().getDriver().getUnderLayingDriver();`

throws 
`java.lang.ClassCastException: com.qmetry.qaf.automation.ui.webdriver.QAFExtendedWebDriver cannot be cast to io.appium.java_client.AppiumDriver`

Because the driver is created via 
`return new QAFExtendedWebDriver(new URL(url), desiredCapabilities, reporter);`

### Expected behavior

Appium **local** driver is created.

### Actual behavior

RemoreDriver is created

### Is the issue reproducible on runner?
[x] Appium
[x] Perfecto

**Comment by Georgy-Ryabov** [comment]
So, I've made a quick workaround:

```
private static WebDriver getDriverObj(Class<? extends WebDriver> of,
			Capabilities capabilities, String urlStr) {
		try {
			if (urlStr.contains("127.0.0.1")) {
				Constructor<? extends WebDriver> constructor =
						of.getConstructor(Capabilities.class);
				return constructor.newInstance(capabilities);
			} else {
				Constructor<? extends WebDriver> constructor =
						of.getConstructor(URL.class, Capabilities.class);
				return constructor.newInstance(new URL(urlStr), capabilities);
			}
		} catch (Exception e) {
			if (e.getCause() != null && e.getCause() instanceof WebDriverException) {
				throw (WebDriverException) e.getCause();
			}
			try {
				return of.newInstance();
			} catch (Exception e1) {
				try {

					Constructor<? extends WebDriver> constructor =
							of.getConstructor(URL.class, Capabilities.class);

					return constructor.newInstance(new URL(urlStr), capabilities);
				} catch (InvocationTargetException e2) {
					throw new WebDriverException(e2);
				} catch (InstantiationException e2) {
					throw new WebDriverException(e2);
				} catch (IllegalAccessException e2) {
					throw new WebDriverException(e2);
				} catch (IllegalArgumentException e2) {
					throw new WebDriverException(e2);
				} catch (MalformedURLException e2) {
					throw new WebDriverException(e2);
				} catch (NoSuchMethodException e2) {
					throw new WebDriverException(e2);
				} catch (SecurityException e2) {
					throw new WebDriverException(e2);
				}
			}
		}
	}
```

And it works for my case

**Comment by cjayswal** [comment]
use driver.name=appuimDriver with capability driverClass io.appium.java_client.ios.IOSDriver or io.appium.java_client.android.AndroidDriver. Refer [How to run on Native application on appium](https://qmetry.github.io/qaf/latest/how_to_run_on_native_app_on_appium.html)

**Comment by Georgy-Ryabov** [comment]
Yes, I used driver.class=io.appium.java_client.ios.IOSDriver

But I have installed appium node js  locally. So, if node js is installed the issue is occurred.
With uninstalled nodeJS there will no any issue with connection to Perfecto.

**Comment by cjayswal** [comment]
Use driverClass capability instead of driver.class. provide perfecto url as remote.server. in that case you don't need appointment server. All you need to have appointment client dependency

**Comment by cjayswal** [comment]
Also make sure you're using latest qaf and providing capability correctly. Refer documentation about how to provide capabilities.

**Comment by Georgy-Ryabov** [comment]
Using driverClass  didn't help to me. I want to develop tests locally with simulator or real device, and then run tests on Perfecto. That's why I have local appium server.

You can test it by installing 
$ npm install -g appium
$ appium

**Comment by cjayswal** [comment]
Are you still facing issue? To run on appium, before execution make sure
1. You followed appium setup instructions 
2. Start appium server 
3. Set remote server and port running appium, for example:
    ```properties
    remote.server=127.0.0.1
    remote.port=4732
   ```
To run on perfecto set remote server to perfecto
```properties
remote.server=https://mycloud.perfectomobile.com/nexperience/perfectomobile/wd/hub
```

**Comment by Georgy-Ryabov** [comment]
`Are you still facing issue? `

Yes, I still have the issue. 

I use one bunch of tests to run it locally and with Perfeco.

So, to run locally I set up appium server and properties as you described: 

```
remote.server=127.0.0.1
remote.port=4732
driver.name=appiumDriver
appium.capabilities.driverClass=io.appium.java_client.ios.IOSDriver
```

That is working fine.


But if I had installed appium server on my machine and run with Perfecto, changed properties as you described, of cause,
```
remote.server=https://mycloud.perfectomobile.com/nexperience/perfectomobile/wd/hub
```

I have got the issue with connection to Perfecto. Because of qaf function getDriverObj  the connection to local appium server is created.


There are two solution to fix it:

1) Remove local appium server if you are using Perfecto
2) Fix the `getDriverObj` function. For instance, do not ignore `urlStr`  parameter
